algorithms/a3c/env_vizdoom.py

```diff
@@ -0,0 +1,379 @@
+        self.curr_file_folder = os.path.dirname(os.path.realpath(__file__))
+
+        # Reading train and test instructions, relative converted to absolute path
+        self.train_instructions = self.get_instr(os.path.join(self.curr_file_folder,
+                                                 self.params.train_instr_file))
+        self.test_instructions = self.get_instr(os.path.join(self.curr_file_folder,
+                                                             self.params.test_instr_file))
+
+        if self.params.use_train_instructions:
+            self.instructions = self.train_instructions
+        else:
+            self.instructions = self.test_instructions
+
+        self.word_to_idx = self.get_word_to_idx()
+        self.objects, self.object_dict = \
+            self.get_all_objects(os.path.join(self.curr_file_folder, self.params.all_instr_file))
+        self.object_sizes = self.read_size_file(os.path.join(self.curr_file_folder,
+                                                             self.params.object_size_file))
+        self.objects_info = self.get_objects_info()
+
+        # extra stuff to make it work with ai2thor A3C
+        self.task = Task()
+        self.observation_space = spaces.Box(low=0, high=255,
+                                            shape=(3, self.params.frame_height,
+                                                   self.params.frame_width),
+                                            dtype=np.uint8)
+        self.action_space = spaces.Discrete(3)
+        self.task.word_to_idx = self.word_to_idx
+
+    def game_init(self):
+        """Starts the doom game engine."""
+        game = DoomGame()
+        self.params.scenario_path = os.path.join(self.curr_file_folder, self.params.scenario_path)
+        game = set_doom_configuration(game, self.params)
+        game.init()
+        self.game = game
+
+    def reset(self):
+        """Starts a new episode.
+        Returns:
+           state: A tuple of screen buffer state and instruction.
+           reward: Reward at that step.
+        """
+
+        self.game.new_episode()
+        self.time = 0
+
+        self.instruction, instruction_id = self.get_random_instruction()
+
+        # Retrieve the possible correct objects for the instruction.
+        correct_objects = self.get_target_objects(instruction_id)
+
+        # Since we fix the number of objects to 5.
+        self.correct_location = np.random.randint(5)
+
+        # Randomly select one correct object from the
+        # list of possible correct objects for the instruction.
+        correct_object_id = np.random.choice(correct_objects)
+        chosen_correct_object = [x for x in self.objects_info if
+                                 x.name == self.objects[correct_object_id]][0]
+
+        # Special code to handle 'largest' and 'smallest' since we need to
+        # compute sizes for those particular instructions.
+        if 'largest' not in self.instruction \
```

Comment:
Ifs should be defined by positives if possible. 
```suggestion
        if 'largest' in self.instruction ...:
            # code currently in else
        else:
           # code currently in if
```

Reply:
I've never heard that but I've always felt that. Did PEP8 or something else tell you this?

Reply:
PEP8 https://www.python.org/dev/peps/pep-0008/#programming-recommendations
```Use is not operator rather than not ... is. While both expressions are functionally identical, the former is more readable and preferred.

Yes:

if foo is not None:
No:

if not foo is None:```

Reply:
you say it as if pep8 was something bad. I think this is a good pattern in general

Reply:
we have to be so careful that this doesn't ruin the logic

Reply:
yeah omg I did this but that is completely different logic just because you had a style change!!!!!! 

These two code samples are not equivalent

```
        if 'largest' not in self.instruction and 'smallest' not in self.instruction:
            object_ids = random.sample([x for x in range(len(self.objects))
                                        if x not in correct_objects], 4)
        else:
            object_ids = self.get_candidate_objects_superlative_instr(
                         chosen_correct_object)
            object_ids = [self.object_dict[x] for x in object_ids]
```

```
        if 'largest' in self.instruction and 'smallest' in self.instruction:
            object_ids = self.get_candidate_objects_superlative_instr(
                chosen_correct_object)
            object_ids = [self.object_dict[x] for x in object_ids]
        else:
            object_ids = random.sample([x for x in range(len(self.objects))
                                        if x not in correct_objects], 4)
```